skrf/network.py

```diff
@@ -2282,7 +2282,15 @@ def get_buffer() -> io.StringIO:
                 output.write('!Data is not renormalized\n')
                 output.write('# {} S {} R\n'.format(ntwk.frequency.unit, form))
             else:
-                output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, r_ref))
+                # Write real reference impedance "a" if possible, complex impedance "(a+0j)"
+                # is not supported by most tools. Also, prefer integers if possible.
+                if r_ref.imag == 0 and int(r_ref.real) == r_ref.real:
+                    output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, int(r_ref.real)))
+                elif r_ref.imag == 0:
+                    output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, r_ref.real))
+                else:
+                    # Write complex impedance "(a+bj)".
+                    output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, r_ref))
```

Comment:
Should eventually complex case emit a warning? Can also be left as is.

Reply:
@mhuser Is there any tool in existence (other than scikit-rf) capable of reading a complex impedance in this form? If not, I'll add a warning message here.

Reply:
@biergaizi 
If there is, I know no examples of such tool, but I have not used them all. It is probably non-standard enough to justify the warning. 😄

Reply:
I think I found the problem. I missed the forest while looking at a tree. 

Complex impedance is not supported to begin with (unless `write_z0` is used, which generates a HFSS-style Touchstone file), and in the beginning of `write_touchstone()`, it actually actively checks whether the imaginary part of the impedance is zero.

    if r_ref is None and write_z0 == False:
        if not equal_z0:
            raise ValueError("Network has unequal port impedances but reference impedance for renormalization 'r_ref' is not specified.")
        if have_complex_ports:
            raise ValueError("Network port impedances are complex but reference impedance for renormalization 'r_ref' is not specified.")
        r_ref = ntwk.z0[0, 0]
    elif r_ref is not None:
        if not npy.isscalar(r_ref):
            raise ValueError('r_ref must be scalar')
        if r_ref.imag != 0:
            raise ValueError('r_ref must be real')
        ntwk.renormalize(r_ref)

In other words, `r_ref` must always be a real number (`assert r_ref.imag == 0`) when we execute:

    output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, r_ref))

The output format complex number is purely a string formatting bug, which should not happen. So, instead of a UserWarning, the complex codepath should never be executed (unless there's a bug in scikit-rf). What we need here is not a warning but as assertion failure.